backend/witness/witness.go

```diff
@@ -199,50 +261,62 @@ func (w *Witness) UnmarshalJSON(data []byte) error {
 	if err := dec.Decode(instance); err != nil {
 		return err
 	}
+	// walk through the public AND secret values
+	missingAssignment := func(name string) error {
+		return fmt.Errorf("missing assignment for %s", name)
+	}
+	{
+		chValues := make(chan any)
 
-	// optimistic approach: first try to unmarshall everything. then only the public part if it fails
-	// note that our instance has leaf type == *fr.Element, so the zero value is nil
-	// and is going to make the newWitness method error since it doesn't accept missing assignments
-	_, err = v.FromAssignment(instance, reflect.PtrTo(typ), false)
-	if err != nil {
-		// try with public only
-		_, err := v.FromAssignment(instance, reflect.PtrTo(typ), true)
-		if err != nil {
-			return err
+		go func() {
+			defer close(chValues)
+			schema.Walk(instance, ptrTyp, func(leaf schema.LeafInfo, tValue reflect.Value) error {
+				if leaf.Visibility == schema.Public {
+					if tValue.IsNil() {
+						return missingAssignment(leaf.FullName())
+					}
+					chValues <- reflect.Indirect(tValue).Interface()
+				}
+				return nil
+			})
+			schema.Walk(instance, ptrTyp, func(leaf schema.LeafInfo, tValue reflect.Value) error {
+				if leaf.Visibility == schema.Secret {
+					if tValue.IsNil() {
+						return missingAssignment(leaf.FullName())
+					}
+					chValues <- reflect.Indirect(tValue).Interface()
+				}
+				return nil
+			})
+
+		}()
+
+		if err := w.Fill(s.NbPublic, s.NbSecret, chValues); err == nil {
+			return nil // we can return; if there is an error, we try the public values only
 		}
-		w.Vector = v
-		return nil
 	}
-	w.Vector = v
-	return nil
-}
 
-func (w *Witness) toAssignment(to interface{}, toLeafType reflect.Type) error {
-	if w.Schema == nil {
-		return errMissingSchema
-	}
-	if w.Vector == nil {
-		return fmt.Errorf("%w: empty witness", ErrInvalidWitness)
-	}
+	{
+		chValues := make(chan any)
```

Comment:
I don't follow. Why do we have to walk again with only public values now?

Reply:
can be skipped now that this is separated; but need to identify before if we errored because of missing assignment in the public or secret witness, and discard any elements we read from the secret witness. Will fix.